zone/task_client_state.cpp

```diff
@@ -0,0 +1,2177 @@
+	if (experience_reward > 0) {
+		client->AddEXP(experience_reward);
+	}
+	if (experience_reward < 0) {
+		uint32 pos_reward = experience_reward * -1;
+		// Minimal Level Based Exp reward Setting is 101 (1% exp at level 1)
+		if (pos_reward > 100 && pos_reward < 25700) {
+			uint8 max_level   = pos_reward / 100;
+			uint8 exp_percent = pos_reward - (max_level * 100);
+			client->AddLevelBasedExp(exp_percent, max_level);
+		}
+	}
+
+	client->SendSound();
+}
+
+bool ClientTaskState::IsTaskActive(int task_id)
+{
+	if (m_active_task.task_id == task_id) {
+		return true;
+	}
+
+	if (m_active_task_count == 0 || task_id == 0) {
+		return false;
+	}
+
+	for (auto &active_quest : m_active_quests) {
+		if (active_quest.task_id == task_id) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+void ClientTaskState::FailTask(Client *client, int task_id)
+{
+	LogTasks(
+		"[FailTask] Failing task for character [{}] task_id [{}] task_count [{}]",
+		client->GetCleanName(),
+		task_id,
+		m_active_task_count
+	);
+
+	if (m_active_task.task_id == task_id) {
+		client->SendTaskFailed(task_id, 0, TaskType::Task);
+		// Remove the task from the client
+		client->CancelTask(0, TaskType::Task);
+		return;
+	}
+
+	// TODO: shared tasks
+
+	if (m_active_task_count == 0) {
+		return;
+	}
+
+	for (int i = 0; i < MAXACTIVEQUESTS; i++) {
+		if (m_active_quests[i].task_id == task_id) {
+			client->SendTaskFailed(m_active_quests[i].task_id, i, TaskType::Quest);
+			// Remove the task from the client
+			client->CancelTask(i, TaskType::Quest);
+			return;
+		}
+	}
+}
+
+// TODO: Shared tasks
+bool ClientTaskState::IsTaskActivityActive(int task_id, int activity_id)
+{
+	LogTasks("[IsTaskActivityActive] task_id [{}] activity_id [{}]", task_id, activity_id);
+
+	// Quick sanity check
+	if (activity_id < 0) {
+		return false;
+	}
+	if (m_active_task_count == 0 && m_active_task.task_id == TASKSLOTEMPTY) {
+		return false;
+	}
+
+	int  active_task_index = -1;
+	auto task_type         = TaskType::Task;
+
+	if (m_active_task.task_id == task_id) {
+		active_task_index = 0;
+	}
+
+	if (active_task_index == -1) {
+		for (int i = 0; i < MAXACTIVEQUESTS; i++) {
+			if (m_active_quests[i].task_id == task_id) {
+				active_task_index = i;
+				task_type         = TaskType::Quest;
+				break;
+			}
+		}
+	}
+
+	// The client does not have this task
+	if (active_task_index == -1) {
+		return false;
+	}
+
+	auto info = GetClientTaskInfo(task_type, active_task_index);
+
+	if (info == nullptr) {
+		return false;
+	}
+
+	TaskInformation *p_task_data = task_manager->m_task_data[info->task_id];
+
+	// The task is invalid
+	if (p_task_data == nullptr) {
+		return false;
+	}
+
+	// The activity_id is out of range
+	if (activity_id >= p_task_data->activity_count) {
+		return false;
+	}
+
+	LogTasks(
+		"[IsTaskActivityActive] (Update) task_id [{}] activity_id [{}] activity_state",
+		task_id,
+		activity_id,
+		info->activity[activity_id].activity_state
```

Comment:
Missing fmt braces here for the 3rd argument

Reply:
Good eye